### ISSUE TITLE
Lowering MHA C++/Python to ATen

### ISSUE THREAD

**Issue by bmon581** [pytorch/pytorch#53836]
## 🚀 Feature
Lowering MHA C++/Python code to ATen.

## Motivation

We are currently working on supporting the ML Compute backend with PyTorch. For that, we would like to override the MHA operation. We rely on a similar way to override operations as XLA. However, currently there does not exist any ATen dispatch for the MHA code. The implementation exists in both:
* C++ API: https://github.com/pytorch/pytorch/blob/master/torch/csrc/api/include/torch/nn/functional/activation.h#L845
* Python API: 
https://github.com/pytorch/pytorch/blob/master/torch/nn/functional.py#L4628

Could we replace the two in favor of a single Default Dispatch key which would be registered in ATen?

## Pitch

Register MHA directly in ATen and call it from both the C++ and Python APIs.

- func: multihead_attention(Tensor self, Tensor keys, Tensor values, Tensor in_projection_weight, Tensor out_projection_weight, int embed_dim, int num_heads,
                                Tensor? key_mask=None, Tensor? attention_mask=None, Tensor? in_projection_bias=None, Tensor? out_pojection_bias=None,
                                Tensor? k_attention_bias=None, Tensor? v_attention_bias=None, int key_dim=0, int val_dim=0, float dropout=0.0, bool has_attention_bias=False, bool add_zero_attention=False) -> Tensor

## Alternatives

If that is not possible, the ML Compute backend will have to create its own MHA layer and expose it as nn.mlc.MultiHeadAttention. This is possible, however, it would decrease the ability of users to toggle between devices.

cc @albanD @mruberry @jbschlosser @walterddr @mikaylagawarecki @zhangguanheng66

**Comment by albanD** [comment]
triage review to discuss the mha oncall.

**Comment by albanD** [comment]
I think this has changed quite a bit since this was opened.
The two main things are:
- we now have sdpa that should cover what you want. It has a proper native implementation that can be overwritten as needed.
- We are moving away from the nn mha abstraction in favor of the sdpa above.

**Comment by janeyx99** [comment]
Closing due to latest comment ^
